aten/src/THCUNN/generic/MultiMarginCriterion.cu

```diff
@@ -54,37 +55,71 @@ void THNN_(MultiMarginCriterion_updateOutput)(
-        THCTensor_(data)(state, output_),
-        THCTensor_(data)(state, input),
-        THCIndexTensor_(data)(state, target),
-        weights ? THCTensor_(data)(state, weights) : NULL,
-        nframe, input->size[1],
-        sizeAverage,
-        margin
-      );
+      THCTensor_(resize1d)(state, output, input->size[0]);
+      if (p == 1)
+      {
+        cunn_MultiMarginCriterion_updateOutput_kernel<1, real, accreal> <<<blocks,threads, 0, THCState_getCurrentStream(state)>>>(
+          THCTensor_(data)(state, output),
+          THCTensor_(data)(state, input),
+          THCIndexTensor_(data)(state, target),
+          weights ? THCTensor_(data)(state, weights) : NULL,
+          nframe, input->size[1],
+          false,
+          margin
+        );
+      }
+      else if (p == 2)
+      {
+        cunn_MultiMarginCriterion_updateOutput_kernel<2, real, accreal> <<<blocks,threads, 0, THCState_getCurrentStream(state)>>>(
+          THCTensor_(data)(state, output),
+          THCTensor_(data)(state, input),
+          THCIndexTensor_(data)(state, target),
+          weights ? THCTensor_(data)(state, weights) : NULL,
+          input->size[0], input->size[1],
+          false,
+          margin
+        );
+      }
+      THCudaCheck(cudaGetLastError());
     }
-    else if (p == 2)
+    else
     {
-      cunn_MultiMarginCriterion_updateOutput_kernel<2, real, accreal> <<<blocks,threads, 0, THCState_getCurrentStream(state)>>>(
-        THCTensor_(data)(state, output_),
-        THCTensor_(data)(state, input),
-        THCIndexTensor_(data)(state, target),
-        weights ? THCTensor_(data)(state, weights) : NULL,
-        input->size[0], input->size[1],
-        sizeAverage,
-        margin
-      );
+      THCTensor_(resize1d)(state, output, 1);
+      THCTensor *output_ = THCTensor_(newWithSize1d)(state, input->size[0]);  // tmp output buffer
+      if (p == 1)
+      {
+        cunn_MultiMarginCriterion_updateOutput_kernel<1, real, accreal> <<<blocks,threads, 0, THCState_getCurrentStream(state)>>>(
+          THCTensor_(data)(state, output_),
+          THCTensor_(data)(state, input),
+          THCIndexTensor_(data)(state, target),
+          weights ? THCTensor_(data)(state, weights) : NULL,
+          nframe, input->size[1],
+          sizeAverage,
+          margin
+        );
+      }
+      else if (p == 2)
+      {
+        cunn_MultiMarginCriterion_updateOutput_kernel<2, real, accreal> <<<blocks,threads, 0, THCState_getCurrentStream(state)>>>(
+          THCTensor_(data)(state, output_),
+          THCTensor_(data)(state, input),
+          THCIndexTensor_(data)(state, target),
+          weights ? THCTensor_(data)(state, weights) : NULL,
+          input->size[0], input->size[1],
+          sizeAverage,
+          margin
+        );
+      }
+      THCudaCheck(cudaGetLastError());
+      float sum = THCTensor_(sumall)(state, output_);
```

Comment:
what's the reason for using a buffer instead of modifying the kernel here?

Reply:
This was the original functionality (use of buffer and then sum afterwards). I just added the non-reduce cases, which don't require using a buffer and summing.